crates/dbsp/src/monitor/mod.rs

```diff
@@ -196,29 +197,25 @@ pub struct TraceMonitorInternal {
 
 impl TraceMonitorInternal {
     fn attach(
-        this: Arc<Mutex<Self>>,
+        this: Rc<RefCell<Self>>,
         circuit: &RootCircuit,
         monitor_eval: bool,
         handler_name: &str,
     ) {
         let this_clone = this.clone();
 
         circuit.register_circuit_event_handler(handler_name, move |event| {
-            let mut this = this.lock().unwrap();
+            let mut this = this.borrow_mut();
             let _ = this.circuit_event(event).map_err(|e| {
                 (this.circuit_error_handler)(event, &e);
             });
         });
         if monitor_eval {
             circuit.register_scheduler_event_handler(handler_name, move |event| {
-                // The `ClockEnd` event can be triggered by `CircuitHandle::drop()` while the
-                // thread is being terminated, at which point the lock may be
-                // poisoned.  Just ignore the event in that case.
-                if let Ok(mut this) = this_clone.lock() {
-                    let _ = this.scheduler_event(event).map_err(|e| {
-                        (this.scheduler_error_handler)(event, &e);
-                    });
-                };
+                let mut this = this_clone.borrow_mut();
```

Comment:
Let's make this `this_clone.try_borrow_mut()` to be on the safe side. While my original comment may not be true, there might be other contexts where the event handler is invoked while holding a reference to the monitor.